lib/AST/Decl.cpp

```diff
@@ -2293,6 +2304,10 @@ CanType ValueDecl::getOverloadSignatureType() const {
                                ->getCanonicalType();
   }
 
+  if (isa<EnumElementDecl>(this)) {
+    return this->getInterfaceType()->getCanonicalType();
+  }
```

Comment:
Note that by giving enum elements different overload types, the overloading of enum cases by associated type becomes legal, e.g allowing:

```swift
enum E {
  case fn(Int)
  case fn(String)
}
```

This is something we ultimately want for [SE-0155], but I don't believe the implementation is currently in place to support it. For now at least, I think the overload type variant of `swift::conflicting` should check to see if it's comparing two enum elements and always consider them to conflict to keep the above example illegal until the implementation is ready.

In addition, I think `mapSignatureFunctionType` should be used here in order to ensure we don't consider things like `@escaping`. For example, for the following:

```swift
enum E {
  case fn(() -> Void)

  static func fn(_ x: @escaping () -> Void) -> E {
    fatalError()
  }
}
```

The method `fn` has the overload type `(E) -> (() -> Void) -> E` (with the `@escaping` having been stripped away by `mapSignatureFunctionType`), whereas the enum element would have the overload type `(E) -> (@escaping () -> Void) -> E`. By using `mapSignatureFunctionType` on the enum element, it would also have its `@escaping` stripped away, bringing the signatures in line.
 


  [SE-0155]: https://github.com/apple/swift-evolution/blob/master/proposals/0155-normalize-enum-case-representation.md

Reply:
Yeah, we should leave overloading enum cases by type (and label) for a separate PR. One place where enum cases appear the mangling is for resilient case index symbols and I need to check if we handle overloading there.